underfs/s3a/src/main/java/alluxio/underfs/s3a/S3ALowLevelOutputStream.java

```diff
@@ -0,0 +1,406 @@
+        execUpload(uploadRequest, mFile, partNumber);
+      }
+
+      waitForAllPartsUpload();
+      completeMultiPartUpload();
+    } catch (Exception e) {
+      LOG.error("Failed to upload {}: {}", mKey, e.toString());
+      throw new IOException(e);
+    }
+    // Set the closed flag, close can be retried until mFile.delete is called successfully
+    mClosed = true;
+  }
+
+  /**
+   * Initializes multipart upload.
+   */
+  private void initMultiPartUpload() throws IOException {
+    // Generate the object metadata by setting server side encryption, md5 checksum,
+    // and encoding as octet stream since no assumptions are made about the file type
+    ObjectMetadata meta = new ObjectMetadata();
+    if (SSE_ENABLED) {
+      meta.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+    }
+    if (mHash != null) {
+      meta.setContentMD5(new String(Base64.encode(mHash.digest())));
+    }
+    meta.setContentType(Mimetypes.MIMETYPE_OCTET_STREAM);
+
+    AmazonClientException lastException;
+    InitiateMultipartUploadRequest initRequest =
+        new InitiateMultipartUploadRequest(mBucketName, mKey).withObjectMetadata(meta);
+    do {
+      try {
+        mUploadId = mClient.initiateMultipartUpload(initRequest).getUploadId();
+        return;
+      } catch (AmazonClientException e) {
+        lastException = e;
+      }
+    } while (mRetryPolicy.attempt());
+    // This point is only reached if the operation failed more
+    // than the allowed retry count
+    throw new IOException("Unable to init multipart upload to " + mKey, lastException);
+  }
+
+  /**
+   * Creates a new temp file to write to.
+   */
+  private void initNewFile() throws IOException {
+    mFile = new File(PathUtils.concatPath(CommonUtils.getTmpDir(), UUID.randomUUID()));
+    if (mHash != null) {
+      mLocalOutputStream =
+          new BufferedOutputStream(new DigestOutputStream(new FileOutputStream(mFile), mHash));
+    } else {
+      mLocalOutputStream = new BufferedOutputStream(new FileOutputStream(mFile));
+    }
+    mOffset = 0;
+    LOG.debug("Init new temp file @ {}", mFile.getPath());
+  }
+
+  /**
+   * Uploads part async.
+   */
+  private void uploadPart() throws IOException {
+    if (mFile == null) {
+      return;
+    }
+    mLocalOutputStream.close();
+    int partNumber = mPartNumber.getAndIncrement();
+    File newFileToUpload = new File(mFile.getPath());
+    mFile = null;
+    mLocalOutputStream = null;
+    UploadPartRequest uploadRequest = new UploadPartRequest()
+        .withBucketName(mBucketName)
+        .withKey(mKey)
+        .withUploadId(mUploadId)
+        .withPartNumber(partNumber)
+        .withFile(newFileToUpload)
+        .withPartSize(newFileToUpload.length());
+    execUpload(uploadRequest, newFileToUpload, partNumber);
+  }
+
+  /**
+   * Executes the upload part request.
+   */
+  private void execUpload(UploadPartRequest request, File file, int partNumber) {
```

Comment:
`file` and `partNumber` can be retrieved from `request`

Reply:
Removed those parameters, thanks